src/IRM/IRMAdaptiveCurve.sol

```diff
@@ -0,0 +1,208 @@
+// SPDX-License-Identifier: MIT
+// Copyright (c) 2023 Morpho Association
+
+pragma solidity ^0.8.0;
+
+import {IIRM} from "evk/InterestRateModels/IIRM.sol";
+import {ExpLib} from "./lib/ExpLib.sol";
+
+/// @title IRMAdaptiveCurve
+/// @custom:contact security@euler.xyz
+/// @author Euler Labs (https://www.eulerlabs.com/).
+/// @author Adapted from Morpho Labs (https://github.com/morpho-org/morpho-blue-irm/).
+/// @notice A Linear Kink IRM that adjusts the rate at target utilization based on time spent above/below it.
+/// @dev This implementation intentionally leaves variables names, units and ExpLib unchanged from original.
+/// Returned rates are extended to RAY per second to be compatible with the EVK.
+contract IRMAdaptiveCurve is IIRM {
+    /// @dev Unit for internal precision.
+    int256 internal constant WAD = 1e18;
+    /// @dev Unit for internal precision.
+    int256 internal constant YEAR = int256(365.2425 days);
+    /// @notice The name of the IRM.
+    string public constant name = "IRMAdaptiveCurve";
+    /// @notice The utilization rate targeted by the model.
+    /// @dev In WAD units.
+    int256 public immutable TARGET_UTILIZATION;
+    /// @notice The initial interest rate at target utilization.
+    /// @dev In WAD per second units.
+    /// When the IRM is initialized for a vault this is the rate at target utilization that is assigned.
+    int256 public immutable INITIAL_RATE_AT_TARGET;
+    /// @notice The minimum interest rate at target utilization that the model can adjust to.
+    /// @dev In WAD per second units.
+    int256 public immutable MIN_RATE_AT_TARGET;
+    /// @notice The maximum interest rate at target utilization that the model can adjust to.
+    /// @dev In WAD per second units.
+    int256 public immutable MAX_RATE_AT_TARGET;
+    /// @notice The steepness of the interest rate line.
+    /// @dev In WAD units.
+    int256 public immutable CURVE_STEEPNESS;
+    /// @notice The speed at which the rate at target is adjusted up or down.
+    /// @dev In WAD per second units.
+    /// For example, with `2e18 / 24 hours` the model will 2x `rateAtTarget` if the vault is fully utilized for a day.
+    int256 public immutable ADJUSTMENT_SPEED;
+
+    /// @notice Internal cached state of the interest rate model.
+    struct IRState {
+        /// @dev The current rate at target utilization.
+        uint208 rateAtTarget;
+        /// @dev The timestamp of the last update to the model.
+        uint48 lastUpdate;
+    }
+
+    /// @notice Get the internal cached state of a vault's irm.
+    mapping(address => IRState) internal irState;
```

Comment:
Can we get an accessor to expose the `lastUpdate` at least? This may be useful to determine if a vault has ever had this IRM installed.